### ISSUE TITLE
Docker eats up 100% RAM when exposing a big range of ports

### ISSUE THREAD

**Issue by Jaykah** [docker/docker#11185]
I have a dockerized SIP application that requires a large number ports for RTP. Since it recently became possible to expose a port range, I decided to move away from --net=host scenario to using those ranges (-p 30000-40000:30000-40000/udp)

However, when a range is large enough, docker eats up all RAM and fails:

```
ERRO[0230] Handler for POST /containers/{name:.*}/start returned error: Cannot start container 80df70ab22d94408e9a5a2c60590b1b1281e5a59b5531590738739c9f7c7c485: iptables failed: iptables --wait -t nat -A DOCKER -p udp -d 0/0 --dport 38207 ! -i docker0 -j DNAT --to-destination 172.17.0.3:38207:  (fork/exec /sbin/iptables: cannot allocate memory)
ERRO[0230] HTTP Error: statusCode=500 Cannot start container 80df70ab22d94408e9a5a2c60590b1b1281e5a59b5531590738739c9f7c7c485: iptables failed: iptables --wait -t nat -A DOCKER -p udp -d 0/0 --dport 38207 ! -i docker0 -j DNAT --to-destination 172.17.0.3:38207:  (fork/exec /sbin/iptables: cannot allocate memory)
```

It would seem logical to combine those ranges when applied to iptables into something like:
--dports 30000:40000, and afaik we do not need to explicitly specify the new destination ports if they're going to match the original ones. Or am I missing something?

**Comment by thaJeztah** [comment]
@brahmaroutu perhaps you have thoughts on this issue, since you implemented this in https://github.com/docker/docker/pull/9097

Also, (possibly) similar, but for `EXPOSE`; https://github.com/docker/docker/issues/9021

**Comment by brahmaroutu** [comment]
@thaJeztah I tried this test and I get the failure
Error response from daemon: Cannot start container 83bc8a32f7a588903a80ca5dfa6c0d699501ff542c7ad8cf365dd95543ffe4a6: Error starting userland proxy: 

But all the ports are mapped when I do an inspect. Looks like it is a timing or memory related issue in the Userland proxy.

**Comment by thaJeztah** [comment]
>Looks like it is a timing or memory related issue in the Userland proxy

Hmf, really hope we can get rid of that once (the proxy).

@brahmaroutu do you think additional info is required on this? Also, perhaps you are able to answer this;

> It would seem logical to combine those ranges when applied to iptables into something like:
--dports 30000:40000, and afaik we do not need to explicitly specify the new destination ports if they're going to match the original ones. Or am I missing something?

**Comment by brahmaroutu** [comment]
@thaJeztah  I agree with @jaykah that only way to avoid running many parallel jobs to allocate all these ports is to block allocate. Not sure refactoring that code, when Docker network drivers are changing, would be a good idea.
The documentation intended to say that if you specify -p 1-10:100-110 then both the ranges should have 10 ports to match. I did not default the case where -p 1-10  would imply -p 1-10:1-10

**Comment by unclejack** [comment]
@Jaykah Can you check how many docker proxy instances are started? I wouldn't be surprised if the root cause was the execution of too many docker proxy processes to handle all of these port mappings.

**Comment by Jaykah** [comment]
@unclejack Hundreds of them, http://prntscr.com/6dj6uu

**Comment by unclejack** [comment]
@Jaykah Thanks!

@crosbymichael @LK4D4 The user land proxy needs to be removed or at least not enabled for port ranges.

**Comment by Jaykah** [comment]
@crosbymichael @LK4D4 @unclejack Hi guys, I understand that the solution to this may take a while, so I was wondering if you could suggest a (temporary) workaround to this issue. If there's a way to disable the proxy that seems like the way to go - unfortunately I'm not familiar enough with Go to do it myself.

Thanks!

**Comment by thaJeztah** [comment]
There's actually some preparations in place to facilitate that (remove the proxy), see https://github.com/docker/docker/pull/11208#issuecomment-77905451, so there's hope :)

**Comment by LK4D4** [comment]
@Jaykah I'm pretty sure that there will be solution in 1.6. For now if you know how to build docker you can replace body of this https://github.com/docker/docker/blob/master/daemon/networkdriver/portmapper/proxy.go#L115 function with return nil. But this will break inter-container communication :(

**Comment by Jaykah** [comment]
Thanks guys! 

@LK4D4 so I assume I have to change that to:

```
func (p *proxyCommand) Start() error {
 return nil;
 r, w, err := os.Pipe()
 if err != nil {
  return fmt.Errorf("proxy unable to open os.Pipe %s", err)
 }
```

Is that correct?

Also, how soon do you think 1.6 will be released?

**Comment by LK4D4** [comment]
@Jaykah 
```
func (p *proxyCommand) Start() error {
    return nil
}
```
It will be beginning of April I think

**Comment by MiLk** [comment]
Same issue here when trying to run https://github.com/QubitProducts/bamboo and listen on 10000-2000.
I had to use `--net=host`.

**Comment by LK4D4** [comment]
@MiLk It spawns whole docker binary on each port :)

**Comment by cpuguy83** [comment]
This might be good to make only 1 docker-proxy per container (if it is publishing ports) instead of per-port.

**Comment by unclejack** [comment]
@cpuguy83 I disagree, the only acceptable and proper solution is the complete removal (disabled and deleted from the code) of the user land proxy.

**Comment by cpuguy83** [comment]
@unclejack We keep talking about that...

**Comment by Jaykah** [comment]
Hey guys. I saw the release of 1.6, any changes to user land proxy in that version?

**Comment by thaJeztah** [comment]
@Jaykah unfortunately, not yet

**Comment by allencloud** [comment]
Hey guys.

Then if I kill the hundreds of docker-proxy processes on my docker daemon machine, will it affect my containers' running?

In addition, I will never use 0.0.0.0:port －> container_ip:port to access container's application. 

@LK4D4 @unclejack @thaJeztah

**Comment by oblop** [comment]
Any news to when the userland docker-proxy will be removed?
No milestone yet for this?
Thanks

**Comment by cpuguy83** [comment]
@oblop It's optionally disabled on master, will be part of 1.7

**Comment by darknigh** [comment]
I also face this problem now. I need to publish a large range of port and it will start a lot of docker-proxy instances. Can anyone tell me how to solve it? I can not use the --net=host because the ssh will not work and some command like passwd will failed.

**Comment by LK4D4** [comment]
@darknigh In 1.7 you will be available to disable it with `--userland-proxy=false`.

**Comment by maxhawkins** [comment]
Did this make it into 1.7? I'm getting "flag not found" when I run with --userland-proxy=false on 1.7.

**Comment by thaJeztah** [comment]
> Did this make it into 1.7? I'm getting "flag not found" when I run with --userland-proxy=false on 1.7.

@maxhawkins yes, it's in 1.7. See: https://github.com/docker/docker/blob/v1.7.0/daemon/config_linux.go#L75. It is an *daemon* option, so should be provided when starting the docker daemon.

**Comment by blop** [comment]
Why is the default not set to false ?
Is it going to be false by default in an upcoming release ?

**Comment by cpuguy83** [comment]
@blop Because it is not supported on RHEL6/CentOS6.
With the announcement that we will no longer support RHEL6 in future releases, I suspect this default will change.

**Comment by Jaykah** [comment]
So, despite the new --userland-proxy=false flag, it is still not production-feasible to use big port ranges due to the fact that we are not block allocating those ports. 

Allocating ports individually means that:

1) We need to deal with open file descriptors, since each rule (for each port) will require one of those
2) The startup of such container will take quite a while (around 10 minutes for 10-15 thousand ports) bacause it will always depend on the speed of provisioning of such individual handles/rules

**Comment by 1N50MN14** [comment]
I'm facing a similar issue here when exposing large range of ports (-p 20000-30000:20000-30000/udp) docker daemon pretty much renders the host machine dead. 

Any tips or how to work around this issue at least until it's resolved?

**Comment by mattwilliamson** [comment]
+1

**Comment by darknigh** [comment]
I solved this problem by using iptables and port forwarding in the host side. The rule added to iptables is below:
iptables -A DOCKER -t nat -p tcp ! -i docker0 --dport 24:65535 -j DNAT --to 172.17.0.1:24-65535
iptables -A DOCKER -p tcp -d 172.17.0.1 ! -i docker0 -o docker0 --dport 24:65535 -j ACCEPT
iptables -A POSTROUTING -t nat -p tcp -s 172.17.0.1 -d 172.17.0.1 --dport 24:65535 -j MASQUERADE

**Comment by thaJeztah** [comment]
ping @mrjana @mavenugo any ideas on https://github.com/docker/docker/issues/11185#issuecomment-129085898 ?

**Comment by deitch** [comment]
Any plans to fix this, at least on the iptables side?

**Comment by thaJeztah** [comment]
ping @aboch @mavenugo perhaps you have some thoughts on this?

**Comment by deitch** [comment]
The nice thing is that we don't need to be smart about calculating "the user opened 2000, 2001, 2002, .... 2100" and convert it into a range. As soon as it is anything more than a few ports, the user will describe it in the yml as a range, which we can just take and insert into iptables as a range. It actually should simplify the code to *remove* expanding a port range.

**Comment by vincentsiu** [comment]
Can anyone confirm whether running the docker daemon with --userland-proxy=false fix the issue?  I try but it does not fix the issue for me - I continue to get the following error message.

*ERROR: Cannot start container dcd5227651790c197835e3f2016f8c747bb748f86e95d6492c75f5e3f83ab47d: failed to create endpoint relaydocker_relay_1 on network bridge: iptables failed: iptables --wait -t nat -A DOCKER -p tcp -d 0/0 --dport 33320 -j DNAT --to-destination 172.17.0.2:30903 ! -i docker0: (fork/exec /sbin/iptables: cannot allocate memory)*

I am not sure how to get @darknigh's method to work.  Can anyone elaborate?

**Comment by wimh** [comment]
@vincentsiu Regarding the example iptables rules from darknigh's at https://github.com/docker/docker/issues/11185#issuecomment-138017981;

You need to adjust the iptables rules to your requirements. So replace `24` ,`65535` and `172.17.0.1` with your port range and ip address of the container.
And of course remove the `-p 30000-31000:30000-31000` from the docker command line. 

You can use 

    iptables -L -n -v

to verify if the rules you created manually look similar to the rules created by docker.

**Comment by vincentsiu** [comment]
@wimh thanks for your advise.  

I am new to Docker so I still haven't quite figure this out.  How (and where) can I apply those iptables rules in the context of docker-machine / docker-compose?  

Regarding my setup, I am using Docker Machine from a ubuntu box to create Docker hosts on AWS.  The container I want to deploy is very simple: just an sshd daemon used for tunneling.  A large range of ports (ie. ~ 1000 ports) needs to be exposed.  There's only 1 container, so Docker-compose is over-kill, but it is used currently nonetheless.

I have run the iptables commands on the Ubuntu box:
```
iptables -A DOCKER -t nat -p tcp ! -i docker0 --dport 8000:30000 -j DNAT --to 172.17.0.4:8000-30000
iptables -A DOCKER -p tcp -d 172.17.0.4 ! -i docker0 -o docker0 --dport 8000:30000 -j ACCEPT
iptables -A POSTROUTING -t nat -p tcp -s 172.17.0.4 -d 172.17.0.4 --dport 8000:30000 -j MASQUERADE
```

```
Chain DOCKER (1 references)
 pkts bytes target     prot opt in     out     source               destination
    0     0 ACCEPT     tcp  --  !docker0 docker0  0.0.0.0/0            172.17.0.4           tcp dpt:10009
    0     0 ACCEPT     tcp  --  !docker0 docker0  0.0.0.0/0            172.17.0.4           tcp dpt:10008
    0     0 ACCEPT     tcp  --  !docker0 docker0  0.0.0.0/0            172.17.0.4           tcp dpt:10007
    0     0 ACCEPT     tcp  --  !docker0 docker0  0.0.0.0/0            172.17.0.4           tcp dpt:10006
    0     0 ACCEPT     tcp  --  !docker0 docker0  0.0.0.0/0            172.17.0.4           tcp dpt:10005
    0     0 ACCEPT     tcp  --  !docker0 docker0  0.0.0.0/0            172.17.0.4           tcp dpt:10004
    0     0 ACCEPT     tcp  --  !docker0 docker0  0.0.0.0/0            172.17.0.4           tcp dpt:10003
    0     0 ACCEPT     tcp  --  !docker0 docker0  0.0.0.0/0            172.17.0.4           tcp dpt:10002
    0     0 ACCEPT     tcp  --  !docker0 docker0  0.0.0.0/0            172.17.0.4           tcp dpt:10001
    1    48 ACCEPT     tcp  --  !docker0 docker0  0.0.0.0/0            172.17.0.4           tcp dpt:10000
    0     0 ACCEPT     tcp  --  !docker0 docker0  0.0.0.0/0            172.17.0.4           tcp dpt:2200
    0     0 ACCEPT     tcp  --  !docker0 docker0  0.0.0.0/0            172.17.0.4           tcp dpts:8000:30000***

```

After all this, I am still hit with the same error.  

**ERROR: Cannot start container fb31227048afdc3270f32e07dfb8550942db22d5708aa76dbe34702f0979c259: failed to create endpoint relaydocker_relay_1 on network bridge: iptables failed: iptables --wait -t nat -A DOCKER -p tcp -d 0/0 --dport 10893 -j DNAT --to-destination 172.17.0.6:10893 ! -i docker0:  (fork/exec /sbin/iptables: cannot allocate memory)**

Any advice?

**Comment by vincentsiu** [comment]
I want to add to the bug report that the docker daemon was startede with --userland-proxy=false.  Docker version is 1.9.1 (build a34a1d5).  I am running this on Ubuntu 14.04.

**Comment by LK4D4** [comment]
ping @mrjana @mavenugo 
Can we do something with this?

**Comment by deitch** [comment]
with userland-proxy you get a mess of iptables rules and processes. Without, you still get a mess of iptables rules. The rules need to be fixed for a range either way, and then userland proxy (if enabled) needs to be fixed for a range as well.

**Comment by visik7** [comment]
Hi. 
Any updates on this? 
M.

**Comment by mavenugo** [comment]
@mrjana has all the details on the userland-proxy issue. Assigned it to him to take a look when he can.

**Comment by radius314** [comment]
:+1:

**Comment by pwaller** [comment]
I'm still looking for a solution for forwarding `localhost:[port]` to `container:[port]` without setting `docker0.route_localnet=1` (unless docker does it for me).

**Comment by CRogers** [comment]
Is there any hope of killing the userland proxy? It's a common trap it seems everyone I work with is falling into (wait for ports to become open before running tests using something like https://github.com/palantir/docker-compose-rule - but with userland proxy the ports are **always** open). I would love to see it gone.

**Comment by pwaller** [comment]
I learned that docker does set `docker0.route_localnet=1` when you specify `--userland-proxy=false`, but it still runs a userland proxy for localhost.

**Comment by xiaods** [comment]
Hi. 
Any updates on this?

**Comment by livesoftplock** [comment]
I agree with this thread.
I have a lot of dockers with port range. It is not optimally have a rule for each port in a range. Docker should change iptable policy to not overload iptables.

We do not use userland proxy because it consume a lot of RAM.

**Comment by xiaods** [comment]
@livesoftplock how do you do on disable userland proxy?

**Comment by livesoftplock** [comment]
@xiaods 

Debian 8.3
We had problem in defaults so we have used /etc/systemd/system/multi-user.target.wants/docker.service

--userland-proxy=false added to ExecStart

**Comment by thaJeztah** [comment]
if you're on docker 1.10 or 1.11, you can also use a [`/etc/docker/daemon.json`](https://docs.docker.com/engine/reference/commandline/daemon/#daemon-configuration-file) file, so that you don't have to create [drop in files](https://docs.docker.com/engine/admin/systemd/#custom-docker-daemon-options) for systemd

**Comment by afr1983** [comment]
+1 on this

Some VOIP softwares require a wide-range of UDP ports to be open (2000+ ports) so the time for docker to build all these individual iptables rules is beyond acceptable.

**Comment by deitch** [comment]
Open a year and more. The irony is, as I showed above, it is a really simple fix. Just open a port range in iptables, at least on that end. One iptables entry for each range, not each port; one userland proxy process for each range, not each port.

**Comment by shayneoAtNorwood** [comment]
Bump! Any update yet?

**Comment by crosbymichael** [comment]
@afr1983 for your usecase, i would suggest `--net host` and just have that specific container have access to the hosts network interfaces.  There will be no proxy, iptables, or bride involved.

This is not ideal for all containers but for specific ones with large port ranges or when latency matters most, it will work very well.

**Comment by enst** [comment]
--net host as a workaround helps. But it's inconvenient in some cases like you may have to put the host IP and/or port range in the container config files. 
And if you have multiple containers which need large port range. They'll have to use different port range to avoid conflict.

**Comment by aboch** [comment]
Regarding the userland proxy memory usage in case of  big port range expose, be aware that #23312 drastically reduced it. Please give it a try with docker 1.12.0 or newer.

**Comment by raszi** [comment]
What I am experiencing is that the docker-proxy is leaking when it is mapping an UDP port at least that is the case with with version 1.11.1.

**Comment by caffeine3** [comment]
I think this issue is not fixed even now.

The below is a sample code.
When this command is executed, the docker daemon becomes hung up.

```
vagrant@dev:~$ docker -v
Docker version 1.12.3, build 6b644ec
vagrant@dev:~$ docker run -ti --rm -p 5000-10000/udp alpine date
```

In addition, when port range is 1000, docker is worked but container launch time is very slow.

```
vagrant@dev:~$ time docker run -ti --rm -p 9000-10000/udp alpine date
Thu Nov 24 14:30:36 UTC 2016

real    0m32.147s
user    0m0.019s
sys     0m0.019s
vagrant@dev:~$ time docker run -ti --rm -p 9900-10000/udp alpine date
Thu Nov 24 14:31:05 UTC 2016

real    0m2.959s
user    0m0.014s
sys     0m0.009s
vagrant@dev:~$ time docker run -ti --rm -p 9990-10000/udp alpine date
Thu Nov 24 14:31:10 UTC 2016

real    0m0.757s
user    0m0.020s
sys     0m0.001s
vagrant@dev:~$ time docker run -ti --rm alpine date
Thu Nov 24 14:31:22 UTC 2016

real    0m0.400s
user    0m0.008s
sys     0m0.011s
```

**Comment by thaJeztah** [comment]
@caffeine3 correct, the issue is still open, and is not resolved. Each published port starts a new userland-proxy, which is the culprit here. If you need to enable a large range of ports, you can try disabling the user land proxy (we tried to make this the default: https://github.com/docker/docker/issues/14856, but there is a kernel bug which may result in issues; https://github.com/docker/docker/issues/5618), or use `--network=host` (https://github.com/docker/docker/issues/11185#issuecomment-245983651) to run in the hosts' network stack.

**Comment by caffeine3** [comment]
@thaJeztah 
I've tried to disable user land proxy, but the same problems are occurred.
And also I'm fear the kernel bugs.

I think a current my solution is to use host network.
It seems to work well.

Thank you for your information.

**Comment by stszap** [comment]
This issue has milestone 1.13.0 and that version was released few weeks ago. Was it forgotten?

**Comment by cpuguy83** [comment]
Working on making `docker-proxy` from a 1-process per port to 1-process per daemon. This should help drastically in exposing large port ranges, though I expect there some other pieces that need tweaking in the engine as well.

**Comment by danielapsmaior** [comment]
Hi, 
I'm trying to open a 10000 udp port range. Disabling userland proxy and creating iptables rules myself have resolved the issue apparently.
Is there a way to run those iptables rules automatically when the container starts? The container IP is dynamic. Is there a way to make routing mesh work with this setup?
Thank you all and I hope there's a real fix soon.
Regards.

**Comment by cpuguy83** [comment]
@danielapsmaior Honestly, opening up this many ports is absolutely insane and you probably want to use macvlan instead.

**Comment by danielapsmaior** [comment]
@cpuguy83 10 thousand ports is "default" in Asterisk... each VOIP call can consume up to 4 RTP ports. I'll have a look in macvlan, thank you for your answer.

**Comment by aboch** [comment]
>  Is there a way to make routing mesh work with this setup?

routing mesh does not require userland-proxy

**Comment by danielapsmaior** [comment]
@aboch but i would have to repeat the iptables config on each node, right?

**Comment by aboch** [comment]
No, docker programs the iptables automatically. You should not need to do it manually.
Maybe I did not understand your first comment.

If you start docker daemon with or without `--userland-proxy=false`, docker will program the iptables for the containers you run with `-p`. If you start  with `--userland-proxy=false`, extra iptables rules are programmed.

If instead you are running your apps as docker service (you mentioned routing-mesh), and you publish the ports in vip mode (default mode), regardless userland-proxy is on or off, it won't be used to make the services ports accessible from outside of the host. Only iptables rules are used in this case.

```
$ docker service create --name 10_ports_service  -p 100-110 busybox top
sx6tmo9ybxbhhno5wj4guhh7u
$ sudo iptables -nvL DOCKER-INGRESS
Chain DOCKER-INGRESS (1 references)
 pkts bytes target     prot opt in     out     source               destination         
    0     0 ACCEPT     tcp  --  *      *       0.0.0.0/0            0.0.0.0/0            tcp dpt:30010
    0     0 ACCEPT     tcp  --  *      *       0.0.0.0/0            0.0.0.0/0            state RELATED,ESTABLISHED tcp spt:30010
    0     0 ACCEPT     tcp  --  *      *       0.0.0.0/0            0.0.0.0/0            tcp dpt:30009
    0     0 ACCEPT     tcp  --  *      *       0.0.0.0/0            0.0.0.0/0            state RELATED,ESTABLISHED tcp spt:30009
    0     0 ACCEPT     tcp  --  *      *       0.0.0.0/0            0.0.0.0/0            tcp dpt:30008
    0     0 ACCEPT     tcp  --  *      *       0.0.0.0/0            0.0.0.0/0            state RELATED,ESTABLISHED tcp spt:30008
    0     0 ACCEPT     tcp  --  *      *       0.0.0.0/0            0.0.0.0/0            tcp dpt:30007
    0     0 ACCEPT     tcp  --  *      *       0.0.0.0/0            0.0.0.0/0            state RELATED,ESTABLISHED tcp spt:30007
    0     0 ACCEPT     tcp  --  *      *       0.0.0.0/0            0.0.0.0/0            tcp dpt:30006
    0     0 ACCEPT     tcp  --  *      *       0.0.0.0/0            0.0.0.0/0            state RELATED,ESTABLISHED tcp spt:30006
    0     0 ACCEPT     tcp  --  *      *       0.0.0.0/0            0.0.0.0/0            tcp dpt:30005
    0     0 ACCEPT     tcp  --  *      *       0.0.0.0/0            0.0.0.0/0            state RELATED,ESTABLISHED tcp spt:30005
    0     0 ACCEPT     tcp  --  *      *       0.0.0.0/0            0.0.0.0/0            tcp dpt:30004
    0     0 ACCEPT     tcp  --  *      *       0.0.0.0/0            0.0.0.0/0            state RELATED,ESTABLISHED tcp spt:30004
    0     0 ACCEPT     tcp  --  *      *       0.0.0.0/0            0.0.0.0/0            tcp dpt:30003
    0     0 ACCEPT     tcp  --  *      *       0.0.0.0/0            0.0.0.0/0            state RELATED,ESTABLISHED tcp spt:30003
    0     0 ACCEPT     tcp  --  *      *       0.0.0.0/0            0.0.0.0/0            tcp dpt:30002
    0     0 ACCEPT     tcp  --  *      *       0.0.0.0/0            0.0.0.0/0            state RELATED,ESTABLISHED tcp spt:30002
    0     0 ACCEPT     tcp  --  *      *       0.0.0.0/0            0.0.0.0/0            tcp dpt:30001
    0     0 ACCEPT     tcp  --  *      *       0.0.0.0/0            0.0.0.0/0            state RELATED,ESTABLISHED tcp spt:30001
    0     0 ACCEPT     tcp  --  *      *       0.0.0.0/0            0.0.0.0/0            tcp dpt:30000
    0     0 ACCEPT     tcp  --  *      *       0.0.0.0/0            0.0.0.0/0            state RELATED,ESTABLISHED tcp spt:30000
    0     0 RETURN     all  --  *      *       0.0.0.0/0            0.0.0.0/0           
$ 
```

If instead you using docker service and publishing the ports in `host` mode, then it will use userland-proxy. But if `--userland-proxy=false`, then again it will add the necessary iptables rules.

**Comment by danielapsmaior** [comment]
@aboch I'm using docker stack deploy with an yml to publish the services in N swarm nodes. One of the services is Asterisk with 10 thousand ports. The load was getting too high with the port mapping, so I did as stated by [ghost](https://github.com/docker/docker/issues/11185#issuecomment-138017981). 

I'm not exposing the port range in yml anymore, I'm doing the iptables rules manually. That's the real issue being discussed on this thread.. docker gets slow using the normal method of port range exposing. But I would like to this iptables config in a script on the stack deploy, when the Asterisk service/container is starting, in all stack nodes...

Maybe the container using the host network is the only way to go... Thank you for your answer.

**Comment by aboch** [comment]
Unfortunately the routing-mesh won't be operational for those ports if you do not specify `-p` in your service creation.
Because the iptables programming is only one part of it, then there is the ipvs programming in the ingress sandbox.

**Comment by matthewh** [comment]
Running with Freeswitch with 2000 open ports and running into the same issues. Adding myself to the list and waiting for a solution.

**Comment by kevzettler** [comment]
What is the recommended workaround when the host is MacOs because `--net=host` is not viable per

https://docs.docker.com/network/host/
>The host networking driver only works on Linux hosts, and is not supported on Docker for Mac, >Docker for Windows, or Docker EE for Windows Server.

I have a container that on MacOs host that I need to expose a large range of ports on and seeing no viable option

**Comment by thaJeztah** [comment]
@djs55 ^^ do you know a workaround for Docker Desktop?

**Comment by djs55** [comment]
For development use-cases where it's only necessary to be able to connect to the ports from the host, then you could try enabling the [experimental SOCKS proxy](https://github.com/docker/for-mac/issues/2670#issuecomment-372365274). This will allow you to connect to the container backend IPs directly.

**Comment by csvan** [comment]
Is this still being worked on?

**Comment by zhaocheng173** [comment]
> > Did this make it into 1.7? I'm getting "flag not found" when I run with --userland-proxy=false on 1.7.
> 
> @maxhawkins yes, it's in 1.7. See: https://github.com/docker/docker/blob/v1.7.0/daemon/config_linux.go#L75. It is an _daemon_ option, so should be provided when starting the docker daemon.

For now -userland-proxy=false should it be turned off at Docker /daemon.json? Any better advice

**Comment by Va1** [comment]
still wondering if there's any solution to this planned to be implemented in Docker.

i assume that Podman or some other implementation might be more efficient in this regard, but i did not test myself yet. i will post here if i do. meanwhile, i'd really appreciate if anyone shares some information about such a prospect. thanks.

**Comment by winkee01** [comment]
does this mean docker-proxy will be removed? I still can see the docker-proxy process work in the background.

**Comment by thaJeztah** [comment]
> does this mean docker-proxy will be removed? I still can see the docker-proxy process work in the background.

Yes, when using `--network=host`, no `docker-proxy` process will be started for that container; here's an example that first creates a container without `--network-host`, which starts a proxy for the published ports (one for IPv4, one for IPv6) then removes the container, and starts a container with `--network=host`, after which no proxy processes are present:

```bash
$ docker run -d -p 80 --name foo nginx:alpine

$ ps aux | grep '[-]proxy'
root      9556  0.0  0.3 1152904 3936 ?        Sl   09:43   0:00 /usr/bin/docker-proxy -proto tcp -host-ip 0.0.0.0 -host-port 49154 -container-ip 172.17.0.2 -container-port 80
root      9561  0.0  0.3 1152904 3900 ?        Sl   09:43   0:00 /usr/bin/docker-proxy -proto tcp -host-ip :: -host-port 49154 -container-ip 172.17.0.2 -container-port 80

$ docker rm -fv foo
$ docker run -d --network=host --name foo nginx:alpine
$ ps aux | grep '[-]proxy'
# (no results)
```

**Comment by polarathene** [comment]
With the summary below in mind, is there any reason to keep this issue open?

8 years and 84 comments contains a fair amount of information that is no longer relevant today. If there is any remaining action to be taken, perhaps a fresh issue is better for that?

- `userland-proxy` isn't relevant to macOS / Windows Docker Desktop AFAIK (_unclear if the memory issue is a problem there, but may warrant it's own issue if it is?_)
- On Linux hosts we can disable `userland-proxy` via `daemon.json`, or use host mode networking.
- `iptables` rule generation regardless of `userland-proxy` setting is still a concern and affects time to provision a container, but is a separate issue to track.

---

From: https://github.com/moby/moby/issues/11185#issuecomment-262802701

> - _Each published port starts a new `userland-proxy`, which is the culprit here._
> - _If you need to enable a large range of ports, you can try disabling the user land proxy (we tried to make this the default: #14856, but there is a kernel bug which may result in issues; #5618)_

Kernel bug doesn't seem to be reproducible anymore AFAIK and has likely been resolved since.

## Summary of comment history

- Originally when reported, there was no supported way to disable the `userland-proxy` AFAIK. That arrived and was briefly defaulted to but became a problem due to unexpected faliures.
- macOS / Windows Docker Desktop apps have been said to no longer use `userland-proxy` as they manage the network of the VM internally differently, that it's only a concern for Linux? (_without the recently introduced Docker Desktop app linux platform support_)
- `userland-proxy: false` or host mode networking (`--network host`) on Linux hosts are said to avoid the issue.
- Large port ranges (such as 10k range) are considered excessive and [alternative network drivers have been advised instead](https://github.com/moby/moby/issues/11185#issuecomment-292293387) where the issue should not be present.
- A related concern is the `iptables` management of each individual port being said to be [configured separately with advice to configure a range in `iptables` instead](https://github.com/moby/moby/issues/11185#issuecomment-297831836). I do not know if there is plans to pursue that, but that sounds like a separate issue to track.
  - [This comment](https://github.com/moby/moby/issues/11185#issuecomment-171379123) highlights that the `iptables` generation concern applies regardless of `userland-proxy` setting, with `userland-proxy: true` differing by creating proxy processes.
  - An example of [`iptables` rules for port range is shown here](https://github.com/moby/moby/issues/11185#issuecomment-138017981).
  - [This comment](https://github.com/moby/moby/issues/11185#issuecomment-129085898) notes each individual port mapping opens a file descriptor and for 10k range may take over 10 minutes to provision. Back at this time, systems may have also had low limits configured (`LimitNOFILE`), which may have also caused problems if reached. Another limit for number of processes may have also been at risk of being hit?
- [Reproduction via CLI](https://github.com/moby/moby/issues/11185#issuecomment-262789331) with output showing time to start container based on size of port range and noting a large enough port range appearing unresponsive. [Some comments](https://github.com/moby/moby/issues/11185#issuecomment-222168407) became focused on time to start a container creating the individual `iptables` port mappings rather than memory usage that this issue was focused on.
- A release was mentioned in the discussion with a [fix that drastically reduced memory usage for large port ranges](https://github.com/moby/moby/issues/11185#issuecomment-246533709).

**Comment by cpuguy83** [comment]
We actually do plan to fix this and finally have people willing to review network changes, so... soon!
Let's leave this open.

/cc @akerouanton
